include/mxnet/lib_api.h

```diff
@@ -18,33 +18,754 @@
+typedef MXReturnValue (*inferType_t)(std::map<std::string, std::string>,
+                                     std::vector<int>&, std::vector<int>&);
+typedef MXReturnValue (*inferShape_t)(std::map<std::string, std::string>,
+                                      std::vector<std::vector<unsigned int> >&,
+                                      std::vector<std::vector<unsigned int> >&);
+typedef MXReturnValue (*mutateInputs_t)(std::map<std::string, std::string>,
+                                      std::vector<int>&);
+typedef MXReturnValue (*createOpState_t)(std::map<std::string, std::string>,
+                                      CustomStatefulOp**);
+
+/*!
+ * \brief Class to hold custom operator registration
+ */
+class CustomOp {
+ public:
+  explicit CustomOp(const char* op_name) : name(op_name),
+    forward(NULL), backward(NULL), parse_attrs(NULL), infer_type(NULL),
+    infer_shape(NULL), mutate_inputs(NULL), create_opstate(NULL) {}
+  ~CustomOp() {}
+  CustomOp& setForward(fcomp_t fcomp) {
+    forward = fcomp;
+    return *this;
+  }
+  CustomOp& setBackward(fcomp_t fcomp) {
+    backward = fcomp;
+    return *this;
+  }
+  CustomOp& setParseAttrs(parseAttrs_t func) {
+    parse_attrs = func;
+    return *this;
+  }
+  CustomOp& setInferType(inferType_t func) {
+    infer_type = func;
+    return *this;
+  }
+  CustomOp& setInferShape(inferShape_t func) {
+    infer_shape = func;
+    return *this;
+  }
+  CustomOp& setMutateInputs(mutateInputs_t func) {
+    mutate_inputs = func;
+    return *this;
+  }
+  CustomOp& setCreateOpState(createOpState_t func) {
+    create_opstate = func;
+    return *this;
+  }
+
+  /*! \brief operator name */
+  const char* name;
+  /*! \brief operator functions */
+  fcomp_t forward;
+  fcomp_t backward;
+  parseAttrs_t parse_attrs;
+  inferType_t infer_type;
+  inferShape_t infer_shape;
+  mutateInputs_t mutate_inputs;
+  createOpState_t create_opstate;
+};
+
+/*!
+ * \brief Registry class to registers things (ops, properties)
+ *        Singleton class
+ */
+template <class T>
+class Registry {
+ public:
+  /*!
+   * \brief get singleton pointer to class
+   * \returns pointer to class
+   */
+  static Registry* get() {
+    static Registry inst;
+    return &inst;
+  }
+  /*!
+   * \brief add a new entry
+   * \returns new object associated with registered name
+   */
+  T& add(const char* name) {
+    T *entry = new T(name);
+    entries.push_back(entry);
+    return *entry;
+  }
+  int size() {
+    return entries.size();
+  }
+  T& get(int idx) {
+    return *(entries[idx]);
+  }
+
+ private:
+  /*! \brief constructor */
+  Registry() {}
+  /*! \brief destructor */
+  ~Registry() {}
+  /*! \brief map of entries in registry */
+  std::vector<T*> entries;
+};
+
+/*!
+ * \brief Macros to help with string concat
+ * Annoyingly, the concat_ and concat macros are necessary to
+ * be able to use __COUNTER__ in an identifier name 
+ */
+#define _STR_CONCAT_(__a, __b) __a ## __b
```

Comment:
Add some prefix to those macros so that there is no conflict

Reply:
change to MX_STR_CONCAT. Also change:
STRINGIFY --> MX_STRINGIFY
_REGISTER_NAME_ --> MX_REGISTER_NAME_
_REGISTER_DEF_ --> MX_REGISTER_DEF_

Leave REGISTER_OP unchanged, otherwise we need to go and change all the examples. This does not conflict with the NNVM_REGISTER_OP macro so it can stay for now.